File: frontend/src/services/inventoryService.ts
import {
  useMutation,
  useQuery,
  useQueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from "@tanstack/react-query";

import type { Unit } from "../types/menu";
import type { Ingredient, StockMovement } from "../types/inventory";
import { apiRequest } from "./httpClient";

interface CreateIngredientPayload {
  name: string;
  unit: Unit;
  min_stock_threshold: string;
  current_stock?: string;
}

interface CreateStockMovementPayload {
  movement_type: "purchase" | "waste" | "adjustment";
  quantity: string;
  notes?: string | null;
}

const INGREDIENTS_QUERY_KEY = ["inventory", "ingredients"] as const;
const ingredientQueryKey = (id: number | null) => ["inventory", "ingredients", id] as const;
const movementsQueryKey = (id: number | null) => ["inventory", "ingredients", id, "movements"] as const;

// Exported, mirrors TABLES_QUERY_KEY/DISHES_QUERY_KEY's cross-file-export
// precedent: both AppShell.tsx's nav badge and AlertsPage.tsx's list independently
// subscribe to inventory.alerts_changed and need to invalidate this same key.
export const ALERTS_QUERY_KEY = ["inventory", "alerts"] as const;

/**
 * Fetches every Ingredient.
 *
 * A plain list read, used for the recipe-line Ingredient picker and the
 * Ingredients screen's own list.
 *
 * @returns The TanStack Query result for the full Ingredient list.
 */
export function useIngredients(): UseQueryResult<Ingredient[], Error> {
  return useQuery({
    queryKey: INGREDIENTS_QUERY_KEY,
    queryFn: () => apiRequest<Ingredient[]>("/api/inventory/ingredients"),
    // Matches authService's deliberate opt-out. The app-level QueryClient sets no
    // retry, so the default of 3 attempts with backoff would turn a 401/403/404
    // into four requests and a multi-second wait before the error state settles.
    retry: false,
  });
}

/**
 * Fetches every Ingredient currently in shortage.
 *
 * @param enabled - Whether the query should run at all. AppShell.tsx calls this
 *   unconditionally (hooks cannot be called conditionally) but only a
 *   warehouse_manager has an Alerts nav item to badge, so it passes `false` for
 *   every other Role rather than firing a request that would only 403.
 *   AlertsPage.tsx, reachable only by a warehouse_manager (route guard), omits
 *   this and always fetches.
 * @returns The TanStack Query result for the derived low-stock alert list.
 */
export function useAlerts(enabled = true): UseQueryResult<Ingredient[], Error> {
  return useQuery({
    queryKey: ALERTS_QUERY_KEY,
    queryFn: () => apiRequest<Ingredient[]>("/api/inventory/alerts"),
    enabled,
    retry: false,
  });
}

/**
 * Creates a new Ingredient.
 *
 * Invalidates ALERTS_QUERY_KEY too, not just INGREDIENTS_QUERY_KEY: a new
 * Ingredient can be created with current_stock already below min_stock_threshold, and nothing
 * else would ever refresh the alerts list for it — record_movement's own crossing-triggered
 * broadcast never fires here, since no Stock Movement was involved. Without this,
 * a newly-created in-shortage Ingredient would render with no shortage styling/sort-to-top on
 * IngredientsPage.tsx until some unrelated event happened to invalidate the cache.
 *
 * @returns The TanStack Query mutation for submitting a new Ingredient.
 */
export function useCreateIngredient(): UseMutationResult<Ingredient, Error, CreateIngredientPayload> {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (payload: CreateIngredientPayload) =>
      apiRequest<Ingredient>("/api/inventory/ingredients", {
        method: "POST",
        body: JSON.stringify(payload),
      }),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: INGREDIENTS_QUERY_KEY });
      void queryClient.invalidateQueries({ queryKey: ALERTS_QUERY_KEY });
    },
  });
}

/**
 * Deactivates an Ingredient, blocking new Recipe Ingredient lines and new
 * Stock Movements against it.
 *
 * Invalidates both INGREDIENTS_QUERY_KEY and ALERTS_QUERY_KEY on settle, matching
 * useCreateIngredient's own double-invalidation: a deactivated Ingredient's is_active flag
 * changed (the ingredients list must show it), and while deactivation itself never changes
 * current_stock, invalidating alerts alongside it keeps this mutation consistent with every
 * other one in this file rather than leaving one silent exception.
 *
 * @returns The TanStack Query mutation for deactivating an Ingredient by id.
 */
export function useDeactivateIngredient(): UseMutationResult<Ingredient, Error, number> {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (ingredientId: number) =>
      apiRequest<Ingredient>(`/api/inventory/ingredients/${ingredientId}/deactivate`, { method: "POST" }),
    onSettled: () => {
      void queryClient.invalidateQueries({ queryKey: INGREDIENTS_QUERY_KEY });
      void queryClient.invalidateQueries({ queryKey: ALERTS_QUERY_KEY });
    },
  });
}

/**
 * Reactivates a previously deactivated Ingredient.
 *
 * @returns The TanStack Query mutation for reactivating an Ingredient by id.
 */
export function useReactivateIngredient(): UseMutationResult<Ingredient, Error, number> {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (ingredientId: number) =>
      apiRequest<Ingredient>(`/api/inventory/ingredients/${ingredientId}/reactivate`, { method: "POST" }),
    onSettled: () => {
      void queryClient.invalidateQueries({ queryKey: INGREDIENTS_QUERY_KEY });
      void queryClient.invalidateQueries({ queryKey: ALERTS_QUERY_KEY });
    },
  });
}

/**
 * Fetches one Ingredient by id, for the Ingredient detail screen's stat cards.
 *
 * @param ingredientId - The id of the Ingredient to fetch, or null while the route
 *   param has not resolved to a usable id yet.
 * @returns The TanStack Query result for this Ingredient.
 */
export function useIngredient(ingredientId: number | null): UseQueryResult<Ingredient, Error> {
  return useQuery({
    queryKey: ingredientQueryKey(ingredientId),
    queryFn: () => apiRequest<Ingredient>(`/api/inventory/ingredients/${ingredientId}`),
    enabled: ingredientId !== null,
    retry: false,
  });
}

/**
 * Fetches every Stock Movement recorded for an Ingredient, newest first.
 *
 * @param ingredientId - The id of the Ingredient whose history is being read, or null
 *   while the route param has not resolved to a usable id yet.
 * @returns The TanStack Query result for this Ingredient's movement history.
 */
export function useStockMovements(ingredientId: number | null): UseQueryResult<StockMovement[], Error> {
  return useQuery({
    queryKey: movementsQueryKey(ingredientId),
    queryFn: () => apiRequest<StockMovement[]>(`/api/inventory/ingredients/${ingredientId}/movements`),
    enabled: ingredientId !== null,
    retry: false,
  });
}

/**
 * Logs a manual Stock Movement against an Ingredient.
 *
 * Invalidates three keys on settle, not just one: `current_stock` changed, which is cached
 * under the single-ingredient key, the movements-list key, *and* the plain ingredients-list
 * key `IngredientsPage.tsx` already reads, so a Warehouse Manager who logs a movement and
 * then navigates back to the list must not see stale stock. `onSettled`, not `onSuccess`,
 * matching `useEditOrderItem`'s reasoning: a rejected submission (422) means nothing changed
 * here, so the extra invalidation is a harmless no-op, but the pattern stays consistent
 * across this codebase's mutations.
 *
 * @param ingredientId - The id of the Ingredient the movement applies to, or null while the
 *   route param has not resolved to a usable id yet.
 * @returns The TanStack Query mutation for submitting a new Stock Movement.
 */
export function useRecordStockMovement(
  ingredientId: number | null,
): UseMutationResult<StockMovement, Error, CreateStockMovementPayload> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: CreateStockMovementPayload) =>
      apiRequest<StockMovement>(`/api/inventory/ingredients/${ingredientId}/movements`, {
        method: "POST",
        body: JSON.stringify(payload),
      }),
    onSettled: () => {
      void queryClient.invalidateQueries({ queryKey: ingredientQueryKey(ingredientId) });
      void queryClient.invalidateQueries({ queryKey: movementsQueryKey(ingredientId) });
      void queryClient.invalidateQueries({ queryKey: INGREDIENTS_QUERY_KEY });
    },
  });
}

File: backend/services/inventory_service.py
from collections.abc import Sequence
from decimal import Decimal
from typing import Any

from sqlalchemy import func, select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from data_models import CreateIngredientRequest, CreateStockMovementRequest, Ingredient, MovementType, StockMovement, User, UserRole
from exceptions import (
    DuplicateIngredientNameError,
    IngredientNotActiveError,
    IngredientNotFoundError,
    InsufficientStockError,
    StockMovementWouldGoNegativeError,
)
from services.realtime_service import RealtimeService


class InventoryService:
    """Creates and manages Ingredient records.

    Config-free apart from realtime_service, so it is registered as a
    container-level Factory with the logger and realtime_service injected.
    Per-request state such as the DB session is passed into each method as an
    argument, never held on the instance, matching UserService's shape.
    """

    def __init__(self, logger: Any, realtime_service: RealtimeService) -> None:
        """Initialize the service.

        Args:
            logger: The loguru logger injected from the container.
            realtime_service: Injected push-notification service, used to
                broadcast inventory.alerts_changed when a Stock Movement
                crosses an Ingredient's shortage threshold in either
                direction (Story 4.2).
        """
        self._logger = logger
        self._realtime_service = realtime_service

    async def list_ingredients(self, db: AsyncSession) -> Sequence[Ingredient]:
        """List every Ingredient.

        No actor argument: a plain unfiltered read has nothing to reject and
        nothing worth auditing, permissions are Role-level only.

        Args:
            db: The active database session.

        Returns:
            Every Ingredient row, in id order.
        """
        result = await db.execute(select(Ingredient).order_by(Ingredient.id))
        return result.scalars().all()

    async def list_alerts(self, db: AsyncSession) -> Sequence[Ingredient]:
        """List every Ingredient currently in shortage (FR-14).

        A Low-Stock Alert is a derived state, not a stored entity (see the
        story's Scope note): an Ingredient is "in shortage" whenever its
        current_stock is strictly below its min_stock_threshold, computed
        fresh on every call. There is nothing to create, dedupe, or clear —
        at most one row per Ingredient exists to begin with, so "at most one
        active alert per Ingredient" and "no manual dismiss" both hold by
        construction.

        Args:
            db: The active database session.

        Returns:
            Every Ingredient currently below its own min_stock_threshold,
            ordered by name.
        """
        result = await db.execute(
            select(Ingredient)
            .where(Ingredient.current_stock < Ingredient.min_stock_threshold)
            .order_by(Ingredient.name)
        )
        return result.scalars().all()

    async def create_ingredient(
        self, db: AsyncSession, actor: User, payload: CreateIngredientRequest
    ) -> Ingredient:
        """Create a new Ingredient record.

        Args:
            db: The active database session.
            actor: The Warehouse Manager or Admin performing the creation,
                used only for logging.
            payload: The submitted name, unit, threshold, and initial stock.

        Returns:
            The newly created Ingredient.

        Raises:
            DuplicateIngredientNameError: If the name already exists,
                compared without regard to case.
        """
        existing = await db.execute(
            select(Ingredient).where(func.lower(Ingredient.name) == payload.name.lower())
        )
        if existing.scalar_one_or_none() is not None:
            self._logger.warning(
                "Ingredient creation rejected by user_id={}: name={} already exists",
                actor.id,
                payload.name,
            )
            raise DuplicateIngredientNameError()

        ingredient = Ingredient(
            name=payload.name,
            unit=payload.unit,
            current_stock=payload.current_stock,
            min_stock_threshold=payload.min_stock_threshold,
        )
        db.add(ingredient)
        try:
            await db.commit()
        except IntegrityError as exc:
            # The check above loses to a concurrent create of the same name.
            # The unique index is the real arbiter, so translate its violation
            # into the same 409 rather than letting it surface as a 500.
            # Logging before rollback, not after: rollback() expires every object
            # bound to this session, actor included, so reading actor.id afterward
            # raises an unhandled MissingGreenlet.
            self._logger.warning(
                "Ingredient creation rejected by user_id={}: name={} already exists (lost the race)",
                actor.id,
                payload.name,
            )
            await db.rollback()
            raise DuplicateIngredientNameError() from exc
        await db.refresh(ingredient)
        self._logger.info(
            "Ingredient created by user_id={}: ingredient_id={} name={} unit={}",
            actor.id,
            ingredient.id,
            ingredient.name,
            ingredient.unit.value,
        )
        return ingredient

    async def get_ingredient(self, db: AsyncSession, ingredient_id: int) -> Ingredient:
        """Fetch one Ingredient by id, for the Ingredient detail screen's stat cards.

        Args:
            db: The active database session.
            ingredient_id: The id of the Ingredient to fetch.

        Returns:
            The matching Ingredient.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
        """
        return await self._get_ingredient(db, ingredient_id)

    async def deactivate_ingredient(self, db: AsyncSession, actor: User, ingredient_id: int) -> Ingredient:
        """Deactivate an active Ingredient, blocking new Recipe Ingredient lines and new Stock
        Movements against it (Story #3/#4).

        Mirrors UserService.deactivate_user exactly: a simple flag flip, idempotent (a no-op on
        an already-inactive Ingredient), no row-locking needed (this isn't a numeric
        read-modify-write like record_movement). Historical Recipe Ingredient lines and Stock
        Movements referencing this Ingredient are untouched — deactivation only flips is_active,
        it never deletes or reassigns the row.

        Args:
            db: The active database session.
            actor: The Warehouse Manager or Admin performing the deactivation, used only for
                logging.
            ingredient_id: The id of the Ingredient to deactivate.

        Returns:
            The deactivated Ingredient.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
        """
        ingredient = await self._get_ingredient(db, ingredient_id)

        if not ingredient.is_active:
            return ingredient

        ingredient.is_active = False
        await db.commit()
        await db.refresh(ingredient)
        self._logger.info(
            "Ingredient deactivated by user_id={}: ingredient_id={}", actor.id, ingredient_id
        )
        return ingredient

    async def reactivate_ingredient(self, db: AsyncSession, actor: User, ingredient_id: int) -> Ingredient:
        """Reactivate a previously deactivated Ingredient, restoring its normal use.

        Mirrors UserService.reactivate_user exactly.

        Args:
            db: The active database session.
            actor: The Warehouse Manager or Admin performing the reactivation, used only for
                logging.
            ingredient_id: The id of the Ingredient to reactivate.

        Returns:
            The reactivated Ingredient.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
        """
        ingredient = await self._get_ingredient(db, ingredient_id)

        if ingredient.is_active:
            return ingredient

        ingredient.is_active = True
        await db.commit()
        await db.refresh(ingredient)
        self._logger.info(
            "Ingredient reactivated by user_id={}: ingredient_id={}", actor.id, ingredient_id
        )
        return ingredient

    async def list_movements(self, db: AsyncSession, ingredient_id: int) -> Sequence[StockMovement]:
        """List every Stock Movement recorded for an Ingredient, newest first.

        Args:
            db: The active database session.
            ingredient_id: The id of the Ingredient whose history is being read.

        Returns:
            Every Stock Movement for this Ingredient, most recent first.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
        """
        await self._get_ingredient(db, ingredient_id)
        result = await db.execute(
            select(StockMovement)
            .where(StockMovement.ingredient_id == ingredient_id)
            .order_by(StockMovement.timestamp.desc(), StockMovement.id.desc())
        )
        return result.scalars().all()

    async def record_movement(
        self, db: AsyncSession, actor: User, ingredient_id: int, payload: CreateStockMovementRequest
    ) -> StockMovement:
        """Log a manual Stock Movement and update the Ingredient's current stock (AC1/AC2).

        Not a guarded UPDATE (trap 18 does not apply here): nothing about the Ingredient's own
        state blocks a movement the way an OrderItem's status blocks an edit. It is still a
        read-modify-write on current_stock, so the read locks the row (mirrors MenuService's
        _lock_dish, trap 9's shape): without the lock, two concurrent movements on the same
        Ingredient would both read the same starting current_stock and the later commit would
        silently discard the earlier delta, even though both StockMovement audit rows would still
        insert correctly, leaving current_stock disagreeing with its own audit trail (NFR-4). Both
        the Ingredient update and the new StockMovement insert commit together in one transaction
        (NFR-4). purchase/waste apply as +/-quantity; adjustment applies payload.quantity as
        already signed. current_stock is rejected cleanly (before any mutation) if the movement
        would drive it negative (AD-16 reversed): manual movements can no longer drive stock below
        zero.

        Story 4.2: broadcasts inventory.alerts_changed to warehouse_manager connections, but only
        when this movement crosses the shortage threshold in either direction (current_stock <
        min_stock_threshold flips), not on every movement. was_low is read for free from the
        Ingredient row _lock_ingredient already loaded, no extra query; is_low still costs the
        same db.refresh() this method already performs for its own return value, not a new
        round-trip added by the crossing check itself.

        Args:
            db: The active database session.
            actor: The Warehouse Manager or Admin logging the movement.
            ingredient_id: The id of the Ingredient the movement applies to.
            payload: The submitted movement type, quantity, and optional note.

        Returns:
            The newly recorded Stock Movement.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
            IngredientNotActiveError: If the Ingredient is currently deactivated (Story #3/#4).
            StockMovementWouldGoNegativeError: If this movement would drive current_stock below
                zero.
        """
        try:
            ingredient = await self._lock_ingredient(db, ingredient_id)
        except IngredientNotFoundError:
            self._logger.warning(
                "Stock movement rejected by user_id={}: ingredient_id={} not found",
                actor.id,
                ingredient_id,
            )
            raise

        if not ingredient.is_active:
            self._logger.warning(
                "Stock movement rejected by user_id={}: ingredient_id={} is deactivated",
                actor.id,
                ingredient_id,
            )
            raise IngredientNotActiveError()

        was_low = ingredient.current_stock < ingredient.min_stock_threshold

        delta = -payload.quantity if payload.movement_type == MovementType.waste else payload.quantity

        # Checked before any mutation (no rollback needed, nothing written yet): AD-16 reversed,
        # a manual movement that would drive current_stock negative is rejected cleanly rather
        # than applied in full past zero.
        if ingredient.current_stock + delta < 0:
            self._logger.warning(
                "Stock movement rejected by user_id={}: ingredient_id={} current_stock={} "
                "delta={} would go negative",
                actor.id,
                ingredient_id,
                ingredient.current_stock,
                delta,
            )
            raise StockMovementWouldGoNegativeError()

        ingredient.current_stock = ingredient.current_stock + delta

        movement = StockMovement(
            ingredient_id=ingredient_id,
            movement_type=payload.movement_type,
            quantity_change=delta,
            performed_by=actor.id,
            notes=payload.notes,
        )
        db.add(movement)
        await db.commit()
        await db.refresh(movement)
        await db.refresh(ingredient)
        self._logger.info(
            "Stock movement recorded by user_id={}: ingredient_id={} type={} quantity_change={} new_stock={}",
            actor.id,
            ingredient_id,
            payload.movement_type.value,
            delta,
            ingredient.current_stock,
        )

        is_low = ingredient.current_stock < ingredient.min_stock_threshold
        if was_low != is_low:
            self._logger.info(
                "Low-stock alert {} for ingredient_id={}: current_stock={} min_stock_threshold={}",
                "activated" if is_low else "cleared",
                ingredient_id,
                ingredient.current_stock,
                ingredient.min_stock_threshold,
            )
            await self._realtime_service.broadcast(
                [UserRole.warehouse_manager],
                "inventory.alerts_changed",
                {"ingredient_id": ingredient_id},
            )

        return movement

    async def apply_consumption(
        self, db: AsyncSession, ingredient_id: int, quantity: Decimal, actor_id: int, order_id: int
    ) -> bool:
        """Deduct a Recipe-driven consumption amount from an Ingredient (FR-13, Story 5.2).

        Reuses record_movement's own row-lock (trap 9) and threshold-crossing shape, but is a
        distinct method rather than a call to record_movement, for two reasons: (1)
        CreateStockMovementRequest explicitly rejects movement_type=consumption as manually
        submittable, so record_movement's payload contract cannot represent this call at all; (2)
        this deduction must be atomic with OrderService.pick_up_item's own OrderItem status UPDATE
        (AD-6, NFR-3) — record_movement commits its own transaction, which would let the stock
        deduction land even if the status UPDATE's guard later failed, or vice versa. So this
        method locks the Ingredient row and stages both the current_stock decrement and the
        StockMovement insert on the given session, but deliberately does not call db.commit() or
        broadcast anything — pick_up_item's own single commit is what makes the OrderItem
        transition and every Ingredient it touches atomic together, and a low-stock broadcast
        fired before that commit would tell a Warehouse Manager's browser to refetch data that
        was never actually committed.

        Args:
            db: The active database session, part of the caller's own transaction.
            ingredient_id: The id of the Ingredient being consumed.
            quantity: The amount to deduct (RecipeIngredient.quantity * OrderItem.quantity),
                always applied as a subtraction. Rejected before any mutation if it would drive
                current_stock negative (AD-16 reversed) — see InsufficientStockError below.
            actor_id: The id of the Cook performing the triggering pick-up, recorded as the
                StockMovement's performed_by.
            order_id: The id of the Order whose item triggered this consumption, recorded as the
                StockMovement's reference_id (FR-13's "referencing the Order").

        Returns:
            True if this deduction crosses the Ingredient's shortage threshold in either
            direction (was_low != is_low), signaling the caller should broadcast
            inventory.alerts_changed after its own commit succeeds; False otherwise.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
            InsufficientStockError: If this deduction would drive current_stock below zero. The
                caller (OrderService.pick_up_item) is responsible for rolling back any earlier
                deduction already staged on this same session in the same loop (AD-6, whole-item
                all-or-nothing) — this method itself only guards its own single deduction.
        """
        ingredient = await self._lock_ingredient(db, ingredient_id)

        was_low = ingredient.current_stock < ingredient.min_stock_threshold

        # Checked before any mutation on this Ingredient (no partial effect for this one line);
        # the caller is responsible for discarding any earlier line's already-staged deduction in
        # the same pick-up (AD-6, whole-item all-or-nothing rejection).
        if ingredient.current_stock - quantity < 0:
            self._logger.warning(
                "Consumption rejected for order_id={}: ingredient_id={} current_stock={} "
                "quantity={} would go negative",
                order_id,
                ingredient_id,
                ingredient.current_stock,
                quantity,
            )
            raise InsufficientStockError()

        ingredient.current_stock = ingredient.current_stock - quantity

        movement = StockMovement(
            ingredient_id=ingredient_id,
            movement_type=MovementType.consumption,
            quantity_change=-quantity,
            reference_id=order_id,
            performed_by=actor_id,
        )
        db.add(movement)

        is_low = ingredient.current_stock < ingredient.min_stock_threshold
        return was_low != is_low

    async def _get_ingredient(self, db: AsyncSession, ingredient_id: int) -> Ingredient:
        """Fetch an Ingredient by id with no row lock, for read-only callers.

        Args:
            db: The active database session.
            ingredient_id: The id of the Ingredient to fetch.

        Returns:
            The matching Ingredient.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
        """
        ingredient = await db.get(Ingredient, ingredient_id)
        if ingredient is None:
            raise IngredientNotFoundError()
        return ingredient

    async def _lock_ingredient(self, db: AsyncSession, ingredient_id: int) -> Ingredient:
        """Lock an Ingredient row for the rest of the transaction and return it.

        record_movement reads current_stock and then writes a new value derived from it.
        Without a lock, two concurrent movements on the same Ingredient both read the same
        starting current_stock and the later commit silently overwrites the earlier delta, even
        though both StockMovement audit rows still insert correctly (mirrors MenuService._lock_dish,
        trap 9's shape: lock the one row every caller contends on, so they serialize instead of
        racing or deadlocking).

        Args:
            db: The active database session.
            ingredient_id: The id of the Ingredient to lock.

        Returns:
            The matching Ingredient, locked for this transaction.

        Raises:
            IngredientNotFoundError: If no Ingredient matches ingredient_id.
        """
        result = await db.execute(select(Ingredient).where(Ingredient.id == ingredient_id).with_for_update())
        ingredient = result.scalar_one_or_none()
        if ingredient is None:
            raise IngredientNotFoundError()
        return ingredient
